### ISSUE TITLE
Deadlock when calling join() on a std::thread handle when running in a Windows DLL

### ISSUE THREAD

**Issue by pdix0n** [rust-lang/rust#84981]
Hi. I'm trying to run Rust on Windows and I'm running into this hang/deadlock when my DLL is repeatedly loaded, export is called, and then DLL unloaded.

My Rust code is as follows:

```rust
#![cfg(windows)]

use winapi::shared::minwindef::*;

#[no_mangle]
#[allow(non_snake_case, unused_variables)]
extern "system" fn DllMain(dll_module: HINSTANCE, call_reason: DWORD, reserved: LPVOID) -> BOOL {

    const DLL_PROCESS_ATTACH: DWORD = 1;
    const DLL_PROCESS_DETACH: DWORD = 0;

    match call_reason {
        DLL_PROCESS_ATTACH => (),
        DLL_PROCESS_DETACH => (),
        _ => ()
    }
    TRUE
}

#[no_mangle]
#[allow(non_snake_case, unused_variables)]
extern "stdcall" fn MyFunc() {
    println!("spawning rust thread");
    let th = std::thread::spawn(|| {
        println!("hello from Rust thread");
    });
    println!("joining rust thread: {:?}", th.thread().id());
    th.join().unwrap();
}
```
With the following Cargo.toml:

```
[package]
name = "MyDLL"
version = "0.1.0"
authors = ["pdix0n <pdix0n>"]
edition = "2018"

[lib]
crate-type = ["cdylib"]

[dependencies]
winapi = { version = "0.3", features = ["minwindef"] }
```

For loading my DLL I use the following code in C:

```c
#include <stdio.h>
#include <Windows.h>

int main(int argc, char **argv)
{
	if (argc < 1) {
		return 1;
	}

	char *dllPath = argv[1];

	for (int i = 0; ; i++) {

		HMODULE h = LoadLibraryA(dllPath);
		if (!h) {
			printf("LoadLibrary failed\n");
			break;
		}

		typedef VOID (__stdcall *MyFuncFN)(VOID);
		MyFuncFN MyFunc = (MyFuncFN)GetProcAddress(h, "MyFunc");
		if (!MyFunc) {
			printf("GetProcAddress failed\n");
			goto LOOP;
		}

		printf("[%d] Running MyFunc\n", i);
		MyFunc();

LOOP:
		FreeLibrary(h);
	}
	system("pause");
    return 0;
}
```

I expected that this will never hang. But after the loop has run for 1085 iterations, looks like Rust will hang trying to join the thread spawned by the standard library.

I'm building both in 32-bit mode with MSVC (both C and Rust), and have tested both stable and nightly. Am I doing something wrong?

**Comment by nagisa** [comment]
This sounds potentially related to thread locals and their slot limits.

> The constant TLS_MINIMUM_AVAILABLE defines the minimum number of TLS indexes available in each process. This minimum is guaranteed to be at least 64 for all systems. The maximum number of indexes per process is 1,088.

from https://docs.microsoft.com/en-us/windows/win32/procthread/thread-local-storage

Nothing Rust can do about this. Consider not printing (which uses TLS under the covers).

**Comment by pdix0n** [comment]
You are right! I just verified that it's a TLS slot issue by running my loader inside windbg. Using the `!tls -1` command I can see that the slots keep accumulating from 0x0000 to 0x043f (matching the 1088 slots) and then the process starts hanging.

I started reading about TLS and found this: https://docs.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library

> When each thread no longer needs to use a TLS index, it must free the memory whose pointer is stored in the TLS slot. When all threads have finished using a TLS index, use the TlsFree function to release the index.

So I guess Rust doesn't properly clean up after itself by calling the `TlsFree` function when its runtime is done executing. I'm not sure I have the skills to fix this problem.

**Comment by pdix0n** [comment]
This workaround works, and now my loader doesn't hang:

```rust
#![cfg(windows)]

use winapi::shared::minwindef::*;
use winapi::um::processthreadsapi::{TlsAlloc, TlsFree};

#[no_mangle]
#[allow(non_snake_case, unused_variables)]
extern "system" fn DllMain(dll_module: HINSTANCE, call_reason: DWORD, reserved: LPVOID) -> BOOL {

    const DLL_PROCESS_ATTACH: DWORD = 1;
    const DLL_PROCESS_DETACH: DWORD = 0;

    match call_reason {
        DLL_PROCESS_ATTACH => (), 
        DLL_PROCESS_DETACH => (), 
        _ => ()
    }   
    TRUE
}

#[no_mangle]
#[allow(non_snake_case, unused_variables)]
extern "stdcall" fn MyFunc() {
    // Get TLS index before Rust thread
    let tls_idx_start = unsafe { TlsAlloc() };

    // Let Rust leak some TLS slots
    println!("spawning rust thread");
    let th = std::thread::spawn(|| {
        println!("hello from Rust thread");
    }); 
    println!("joining rust thread: {:?}", th.thread().id());
    th.join().unwrap();

    // Get TLS index after Rust thread
    let tls_idx_end = unsafe { TlsAlloc() };

    // Free all TLS slots between tls_idx_start and tls_idx_end
    for tls_idx in tls_idx_start..tls_idx_end+1 {
        unsafe { TlsFree(tls_idx) };
    }   
}
```

Would be cool if this could be fixed in Rust itself though.

**Comment by pdix0n** [comment]
I tried doing some long-running tests with a DLL built using the toolchain with the fix I proposed in #85145. My process didn't hang and the TLS slots were freed, but I noticed a slow continuous increase in memory - possibly related to the TLS memory/objects not being freed. I guess I will just write my code in C instead of this shit language.

**Comment by m-ou-se** [comment]
> possibly related to the TLS memory/objects not being freed.

Yes, that cleanup function would also need to deallocate the `Node` structures next to freeing the TLS slots themselves. This code was originally not designed with freeing these slots in mind: https://github.com/rust-lang/rust/blob/49ca3d9796030fc0a85089460e9f825ceecc08ed/library/std/src/sys/windows/thread_local_key.rs#L87-L92

---

> I guess I will just write my code in C instead of this shit language.

Well that's a very unnecessary comment. What did you hope to achieve with it?

**Comment by pdix0n** [comment]
> Well that's a very unnecessary comment. What did you hope to achieve with it?

I just wanted to let other developers know that I "solved" this issue by coding in C instead, and that Rust isn't really ready for production in this case. The fact that Rust is a shit language might be an unnecessary to you, since you already knew
